File: src/main/java/com/scottscmo/song/adapters/SongSlideTextAdapter.java
package com.scottscmo.song.adapters;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import com.scottscmo.song.Song;

public class SongSlideTextAdapter {
    private final static String DELIM = "\n\n---\n\n";

    public static List<String> serializeToList(Song song, List<String> langs, int maxLines) {
        if (song == null || langs == null || langs.isEmpty()) {
            return Collections.emptyList();
        }

        List<String> slides = new ArrayList<String>();

        for (String verseNumber : song.verseOrder()) {
            Map<String, String[]> verseText = song.getVerseText(verseNumber);

            assert(langs.stream().allMatch(lang -> verseText.containsKey(lang)));

            // assuming all langs have same # of verse lines
            int numVerseLines = verseText.get(langs.get(0)).length;

            int numSlidePerThisVerse = 0;
            while (numSlidePerThisVerse * maxLines < numVerseLines) {
                String slide = "";
                for (String lang : langs) {
                    String[] verseLines = verseText.get(lang);
                    if (verseLines == null) {
                        continue;
                    }
                    for (int i = 0; i < maxLines; i++) {
                        int currLineInVerse = numSlidePerThisVerse * maxLines + i;
                        if (currLineInVerse < verseLines.length) {
                            slide += verseLines[currLineInVerse] + "\n";
                        }
                    }
                }
                slides.add(slide.trim());
                numSlidePerThisVerse++;
            }
        }

        return slides;
    }

    public static String serialize(Song song, List<String> langs, int maxLines) {
        return serializeToList(song, langs, maxLines).stream()
            .collect(Collectors.joining(DELIM));
    }
}

File: src/main/kotlin/com/scottscmo/song/adapters/SongSlideTextAdapter.kt
package com.scottscmo.song.adapters

import com.scottscmo.song.Song
import java.util.stream.Collectors

object SongSlideTextAdapter {
    private const val SEPARATOR = "\n\n---\n\n"

    fun serializeToList(song: Song?, langs: List<String>?, maxLines: Int): List<String> {
        if (song == null || langs == null || langs.isEmpty()) {
            return emptyList()
        }
        val slides: MutableList<String> = ArrayList()
        for (verseNumber in song.verseOrder) {
            val verseText = song.getVerseText(verseNumber)
            assert(verseText != null && langs.all { verseText.containsKey(it) })

            // assuming all langs have same # of verse lines
            val numVerseLines = verseText!![langs[0]]!!.size
            var numSlidePerThisVerse = 0
            while (numSlidePerThisVerse * maxLines < numVerseLines) {
                var slide = ""
                for (lang in langs) {
                    val verseLines = verseText[lang] ?: continue
                    for (i in 0 until maxLines) {
                        val currLineInVerse = numSlidePerThisVerse * maxLines + i
                        if (currLineInVerse < verseLines.size) {
                            slide += verseLines[currLineInVerse] + "\n"
                        }
                    }
                }
                slides.add(slide.trim())
                numSlidePerThisVerse++
            }
        }
        return slides
    }

    fun serialize(song: Song?, langs: List<String>?, maxLines: Int): String {
        return serializeToList(song, langs, maxLines).joinToString { SEPARATOR }
    }
}